### PR TITLE
[dynamo] Forward OptimizedModule.__setattr__ to the wrapped module

### LINKED ISSUE #114844: CUDA graph doesn't update tensor on replay
### 🐛 Describe the bug

CUDA graph results in wrong value after replay for the following example. The test case expects the "x" tensor in the model to be updated with the first linear output, but graph replay doesn't update it.

```
import torch
import os

class Model(torch.nn.Module):
    def __init__(self, inp_size, out_size, inner_size):
        super(Model, self).__init__()
        self.Linear1 = torch.nn.Linear(inp_size, inner_size)
        self.Linear2 = torch.nn.Linear(inner_size, out_size)
        self.x = None

    def forward(self, inp):
        res = self.Linear1(inp)
        self.x = res
        return self.Linear2(res)

N, D_in, H, D_out, inner = 2, 2, 2, 2, 4
model = Model(D_in, H, inner).to("cuda")
input = torch.ones(N, D_in, device='cuda')
out = model(input)

g = torch.cuda.CUDAGraph()
with torch.cuda.graph(g):
  out = model(input)

input2 = torch.ones(N, D_in, device='cuda').add_(10)
# Keep some intermediate value in model.x
model.x = torch.Tensor([[100, 100, 100, 100],[200, 200, 200, 200]])

# Incorrect behavior
input.copy_(input2)
g.replay()
print(f"Incorrect model.x = {model.x}")

# Expected behavior
out = model(input2)
print(f"Expected model.x = {model.x}")
```

The result from the above code -
```
Incorrect model.x = tensor([[100., 100., 100., 100.],
        [200., 200., 200., 200.]])
Expected model.x = tensor([[-10.6180,  -0.6991,  -4.5837,   8.1134],
        [-10.6180,  -0.6991,  -4.5837,   8.1134]], device='cuda:0',
       grad_fn=<AddmmBackward0>)
```
Is this because CUDA graph cannot capture the CPU side "self.x = res" execution and there is no replay for this step? Since this silently gives a wrong value, is there anyway for CUDA graph to detect it?

### Versions

PyTorch version: 2.1.0+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 14.0.0-1ubuntu1.1
CMake version: version 3.27.7
Libc version: glibc-2.35

Python version: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-5.15.120+-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: Tesla T4
Nvidia driver version: 525.105.17
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.6
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.6
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          2
On-line CPU(s) list:             0,1
Vendor ID:                       GenuineIntel
Model name:                      Intel(R) Xeon(R) CPU @ 2.00GHz
CPU family:                      6
Model:                           85
Thread(s) per core:              2
Core(s) per socket:              1
Socket(s):                       1
Stepping:                        3
BogoMIPS:                        4000.30
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat md_clear arch_capabilities
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       32 KiB (1 instance)
L1i cache:                       32 KiB (1 instance)
L2 cache:                        1 MiB (1 instance)
L3 cache:                        38.5 MiB (1 instance)
NUMA node(s):                    1
NUMA node0 CPU(s):               0,1
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Mitigation; PTE Inversion
Vulnerability Mds:               Vulnerable; SMT Host state unknown
Vulnerability Meltdown:          Vulnerable
Vulnerability Mmio stale data:   Vulnerable
Vulnerability Retbleed:          Vulnerable
Vulnerability Spec store bypass: Vulnerable
Vulnerability Spectre v1:        Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:        Vulnerable, IBPB: disabled, STIBP: disabled, PBRSB-eIBRS: Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Vulnerable

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] torch==2.1.0+cu118
[pip3] torchaudio==2.1.0+cu118
[pip3] torchdata==0.7.0
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.16.0
[pip3] torchvision==0.16.0+cu118
[pip3] triton==2.1.0
[conda] Could not collect

cc @mcarilli @ezyang @eellison @peterbell10 @msaroufim @bdhirsh @anijain2305 @zou3519 @chauhang

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122098

Fixes #114844

In the linked issue we have
```
compiled_module = torch.compile(module)
compiled_module.x = ...
compiled_module(...)  # Mutates self.x
```
Where since the module mutates `self.x` you would expect `compiled_module.x`
to be updated but actually `compiled_module.x = ...` sets an attribute "x"
on the `OptimizedModule` object while the forward method of the module mutates
`module.x`.

This gives the expected behavior by forwarding `compiled_module.__setattr__`
down to `module.__setattr__`. There is already a corresponding `__getattr__`
so now `compiled_module.x` becomes an alias for `module.x`.

cc @ezyang @gchanan @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

### CODE REVIEW

File: torch/_dynamo/debug_utils.py [hunk 1]
```diff
@@ -321,19 +319,9 @@ def run_fwd_maybe_bwd(gm, args, only_fwd=False, disable_clone=False):
     if hasattr(gm, "zero_grad"):
         gm.zero_grad(True)
 
-    # TorchInductor returned callable expects lists. So, boxing the call.
-    orig_named_parameters = getattr(gm, "named_parameters", None)
-    orig_named_buffers = getattr(gm, "named_buffers", None)
-    if not hasattr(gm, "_boxed_call") and (
-        orig_named_parameters is not None or orig_named_buffers is not None
-    ):
-        gm = make_boxed_func(gm)
```
Comment (peterbell10) [inline torch/_dynamo/debug_utils.py:330]: I removed the `make_boxed_func` wrapper here as it was interfering with the `collect_results` call below which does `isinstance(gm, OptimizedModule)`.

File: torch/_dynamo/debug_utils.py [hunk 2]
```diff
@@ -359,21 +347,8 @@ def same_two_models(
...
-    )
     from .utils import same
 
-    if isinstance(gm, OptimizedModule):
-        gm.named_parameters = named_parameters_for_optimized_module(gm)
-        gm.named_buffers = named_buffers_for_optimized_module(gm)
-
-    if isinstance(opt_gm, OptimizedModule):
-        opt_gm.named_parameters = named_parameters_for_optimized_module(opt_gm)
-        opt_gm.named_buffers = named_buffers_for_optimized_module(opt_gm)
```
Comment (peterbell10) [inline torch/_dynamo/debug_utils.py:375]: This is aliasing `gm.named_parameters` and `gm.named_buffers` to the underlying `gm._orig_mod` attributes,  which is no longer necessary.

**Review by ezyang — APPROVED** [review]
very nice

**Review by lezcano — APPROVED** [review]

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting your change but the distributed failure looks legit as it is also failing in trunk https://hud.pytorch.org/pytorch/pytorch/commit/88ebdbc97c103271766203df6662240e95a09b42' -c ignoredsignal 

```
2024-03-18T19:44:20.8615564Z FAILED [6.3146s] distributed/checkpoint/test_state_dict.py::TestStateDict::test_fsdp2 - RuntimeError: Process 0 exited with error code 10 and exception:
2024-03-18T19:44:20.8615744Z Traceback (most recent call last):
2024-03-18T19:44:20.8616656Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/common_distributed.py", line 656, in run_test
2024-03-18T19:44:20.8616811Z     getattr(self, test_name)()
2024-03-18T19:44:20.8617718Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/common_distributed.py", line 542, in wrapper
2024-03-18T19:44:20.8617843Z     fn()
2024-03-18T19:44:20.8618693Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/common_utils.py", line 2739, in wrapper
2024-03-18T19:44:20.8618858Z     method(*args, **kwargs)
2024-03-18T19:44:20.8619863Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/distributed/_tensor/common_dtensor.py", line 374, in wrapper
2024-03-18T19:44:20.8620109Z     func(self, *args, **kwargs)  # type: ignore[misc]
2024-03-18T19:44:20.8620995Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/common_distributed.py", line 181, in wrapper
2024-03-18T19:44:20.8621154Z     return func(*args, **kwargs)
2024-03-18T19:44:20.8621843Z   File "/var/lib/jenkins/workspace/test/distributed/checkpoint/test_state_dict.py", line 264, in test_fsdp2
2024-03-18T19:44:20.8621985Z     self.run_subtests(
2024-03-18T19:44:20.8623138Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/distributed/_tensor/common_dtensor.py", line 353, in run_subtests
2024-03-18T19:44:20.8623374Z     return run_subtests(self, *args, **kwargs)
2024-03-18T19:44:20.8624424Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/testing/_internal/distributed/_tensor/common_dtensor.py", line 408, in run_subtests
2024-03-18T19:44:20.8624681Z     test_fn(*test_args, **test_kwargs, **subtest_kwargs)
2024-03-18T19:44:20.8625374Z   File "/var/lib/jenkins/workspace/test/distributed/checkpoint/test_state_dict.py", line 259, in _test_fsdp2
2024-03-18T19:44:20.8625557Z     self._test_save_load(init_model_optim)
2024-03-18T19:44:20.8626275Z   File "/var/lib/jenkins/workspace/test/distributed/checkpoint/test_state_dict.py", line 140, in _test_save_load
2024-03-18T19:44:20.8626447Z     dist_msd = dist_model.state_dict()
2024-03-18T19:44:20.8627252Z   File "/opt/conda/envs/py_3.10/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1911, in state_dict
2024-03-18T19:44:20.8627854Z     module.state_dict(destination=destination, prefix=prefix + name + '.', keep_vars=keep_vars)
2024-03-18T19:44:20.8628543Z TypeError: _patch_model_state_dict.<locals>.state_dict_call() got an unexpected keyword argument 'destination'
```

**Comment by peterbell10** [comment]
@atalman can you share any details about the failure?

**Comment by ezyang** [comment]
The test failure looks like this

```
ERROR: test_dynamo_walk_module (caffe2.torch.fb.module_factory.sync_sgd.tests.test_torchdynamo_wrapper.TorchDynamoWrapperTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/b7e860e5780369a8/caffe2/torch/fb/module_factory/sync_sgd/tests/__test_torchdynamo_wrapper__/test_torchdynamo_wrapper#link-tree/caffe2/torch/fb/module_factory/sync_sgd/tests/test_torchdynamo_wrapper.py", line 163, in test_dynamo_walk_module
    self.assertEqual({x for x, y in params}, expected_params)
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/b7e860e5780369a8/caffe2/torch/fb/module_factory/sync_sgd/tests/__test_torchdynamo_wrapper__/test_torchdynamo_wrapper#link-tree/caffe2/torch/fb/module_factory/sync_sgd/tests/test_torchdynamo_wrapper.py", line 163, in <setcomp>
    self.assertEqual({x for x, y in params}, expected_params)
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/b7e860e5780369a8/caffe2/torch/fb/module_factory/sync_sgd/tests/__test_torchdynamo_wrapper__/test_torchdynamo_wrapper#link-tree/caffe2/torch/fb/training_toolkit/common/proxy_module_thrift/module_walk.py", line 377, in module_walk_parameters
    yield from cast(
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/b7e860e5780369a8/caffe2/torch/fb/module_factory/sync_sgd/tests/__test_torchdynamo_wrapper__/test_torchdynamo_wrapper#link-tree/caffe2/torch/fb/training_toolkit/common/proxy_module_thrift/module_walk.py", line 234, in module_walk
    for module_name, module in _walk_proxy_modules(model):
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/b7e860e5780369a8/caffe2/torch/fb/module_factory/sync_sgd/tests/__test_torchdynamo_wrapper__/test_torchdynamo_wrapper#link-tree/caffe2/torch/fb/training_toolkit/common/proxy_module_thrift/module_walk.py", line 88, in _walk_proxy_modules
    yield from _walk_proxy_modules(child, f"{prefix}{child_name}")
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/b7e860e5780369a8/caffe2/torch/fb/module_factory/sync_sgd/tests/__test_torchdynamo_wrapper__/test_torchdynamo_wrapper#link-tree/caffe2/torch/fb/training_toolkit/common/proxy_module_thrift/module_walk.py", line 85, in _walk_proxy_modules
    yield from _walk_proxy_modules(child, new_prefix)
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/b7e860e5780369a8/caffe2/torch/fb/module_factory/sync_sgd/tests/__test_torchdynamo_wrapper__/test_torchdynamo_wrapper#link-tree/caffe2/torch/fb/training_toolkit/common/proxy_module_thrift/module_walk.py", line 78, in _walk_proxy_modules
    _, child = get_descendant_module_from_str(module, child_path)
  File "/data/sandcastle/boxes/eden-trunk-hg-full-fbsource/buck-out/v2/gen/fbcode/b7e860e5780369a8/caffe2/torch/fb/module_factory/sync_sgd/tests/__test_torchdynamo_wrapper__/test_torchdynamo_wrapper#link-tree/caffe2/torch/fb/training_toolkit/common/proxy_module_thrift/module_walk.py", line 167, in get_descendant_module_from_str
    module = module._modules[child_name]
KeyError: '_orig_mod'

```

**Comment by ezyang** [comment]
The code performing the walking is this:

```
def _walk_proxy_modules(
    module: nn.Module,
    prefix: str = "",
    custom_named_children_fn: Optional[str] = CUSTOM_NAMED_CHILDREN_FN,
) -> Iterator[Tuple[str, nn.Module]]:
    r"""
    This traversal function walks through all the submodules of the given model
    and returns:
    1.) all the modules that have no proxy module as its ancestor,
    2.) all the proxy modules in the model.

    It can also take a custom_named_children_fn to override module.named_children, in
    order to customize the behavior of accessing immediate children.
    custom_named_children_fn on module is expected to have the signature () -> Dict[str, str],
    where the key is a path to descendant, and value is the new name given to it.
    """
    yield prefix, module
    if not ProxyModule.is_proxy_module(module) and not isinstance(
        module, TorchRecShardedModule
    ):
        # No need to walk into proxy module or torchrec's sharded module
        prefix = f"{prefix}." if prefix else ""

        # Some modules like FSDP overloads getattr() and gets attr from the children.
        # Check with dir() to make sure it indeed has the attr.
        if (
            custom_named_children_fn is not None
            and hasattr(module, custom_named_children_fn)
            and custom_named_children_fn in dir(module)
        ):
            named_children_fn = getattr(module, custom_named_children_fn)
            for child_path, child_name in named_children_fn().items():
                _, child = get_descendant_module_from_str(module, child_path)
                if not child_name:
                    # if child_name is empty, it means we want to skip its name,
                    # so we use the original prefix passed in and remove the trailing dot
                    new_prefix = prefix[:-1]
                else:
                    new_prefix = f"{prefix}{child_name}"
                yield from _walk_proxy_modules(child, new_prefix)
        else:
            for child_name, child in module.named_children():
                yield from _walk_proxy_modules(child, f"{prefix}{child_name}")
```

**Comment by ezyang** [comment]
It is likely that `named_children_fn` is this `optimized_child.named_children_walk = lambda: {"_orig_mod": ""}`

So now the problem makes sense: the internal test failure appears to be assuming that an OptimizedModule has an `_orig_mod` attribute which is a submodule. But now orig_mod is no longer being registered to the _modules dictionary for some reason (maybe because setattr is supposed to do it, and your code changed this), so now it fails.

I hope this is enough to debug the rest.

**Comment by peterbell10** [comment]
> the internal test failure appears to be assuming that an `OptimizedModule` has an `_orig_mod` attribute which is a submodule. But now `orig_mod` is no longer being registered to the `_modules` dictionary for some reason 

This doesn't seem to make sense with what I see locally, as the compiled module does have an `_orig_mod` attribute and it's listed in `_modules`:
```python
import torch

mod = torch.nn.BatchNorm2d(10, device="cuda")
a = torch.randn([1, 10, 5, 5], device="cuda")
cmod = torch.compile(mod)

print(f"{=:type(cmod)}")           # <class 'torch._dynamo.eval_frame.OptimizedModule'>
print(list(cmod.named_children())) # [('_orig_mod', BatchNorm2d(...))]
print(cmod._modules)               # OrderedDict([('_orig_mod', BatchNorm2d(...))])
print(mod._modules)                # OrderedDict()
```

which is the same result as before this PR.

> It is likely that `named_children_fn` is this `optimized_child.named_children_walk = lambda: {"_orig_mod": ""}`

If this is true, then the error does make sense as this would set the `named_children_walk` function on `mod` rather than `cmod`. So the walk would expect there to be a `mod._orig_mod`.

I don't really see how we could work around this, so long as `cmod` is pretending to be `mod` then monkey-patching like this isn't going to work quite as expected. Though the monkey-patching could be done by setting `cmod.__dict__` directly.

I'm not sure in context though, is this custom walk function some kind of workaround for an issue with walking optimized modules which is fixed in this PR or is there some other reason not to use `cmod.named_children()`?

**Comment by ezyang** [comment]
I have no idea what the code here does or why it is written this way haha. If you think that we cannot first principles fix the BC break, let's patch the code. Maybe if we add `named_children_walk` to the not pass through list?

**Comment by ezyang** [comment]
It fixes the test.

```
diff --git a/fbcode/caffe2/torch/_dynamo/eval_frame.py b/fbcode/caffe2/torch/_dynamo/eval_frame.py
--- a/fbcode/caffe2/torch/_dynamo/eval_frame.py
+++ b/fbcode/caffe2/torch/_dynamo/eval_frame.py
@@ -125,6 +125,7 @@
         "forward",
         "_forward",
         "__dict__",
+        "named_children_walk",
     }
 
     def __init__(self, mod: torch.nn.Module, dynamo_ctx):
```

**Pull request closed without merging**